### DISCUSSION TITLE
有关bitcast和getelementptr inbounds的问题

### DISCUSSION THREAD

**Discussion by Yawningx** [arcsysu/SYsU-lang#111] [Q&A]
测试tester/function_test2020/15_array_test3.sysu.c样例时，我遇到报错如下：
![8aae1b0c9899d4afa55669054b3e6e4](https://github.com/arcsysu/SYsU-lang/assets/74655243/c748d10d-482d-4f00-9216-5fb88f4f32d0)

标准clang输出为：
%call = call i32 @func([5 x i32]* getelementptr inbounds ([5 x [5 x i32]], [5 x [5 x i32]]* @a, i64 0, i64 0))

想知道可能是什么原因导致了bitcast的出现，而不是gep inbounds？

**Comment by zejia-lin** [comment]
是你的代码产生的bitcast吗

**Reply by Yawningx** [reply]
是的，贴的图是我生成的IR

我生成的main函数IR：
```
define i32 @main() {
entry:
  %0 = call i32 @func([5 x i32*] bitcast ([5 x [5 x i32]]* @"0x644d18" to [5 x i32*]))
  call void @_sysy_putint(i32 %0)
  ret i32 0
}
```
这里的变量@"0x644d18"在全局变量中定义：
@"0x644d18" = global [5 x [5 x i32]] [[5 x i32] [i32 1, i32 2, i32 3, i32 4, i32 5], [5 x i32] zeroinitializer, [5 x i32] zeroinitializer, [5 x i32] zeroinitializer, [5 x i32] zeroinitializer]

clang输出的main函数对应IR：
```
define i32 @main() #0 {
entry:
  %retval = alloca i32, align 4
  store i32 0, i32* %retval, align 4
  %call = call i32 @func([5 x i32]* getelementptr inbounds ([5 x [5 x i32]], [5 x [5 x i32]]* @a, i64 0, i64 0))
  call void @_sysy_putint(i32 %call)
  ret i32 0
}
```
他的变量@a在全局变量中定义如下：
@a = global [5 x [5 x i32]] [[5 x i32] [i32 1, i32 2, i32 3, i32 4, i32 5], [5 x i32] zeroinitializer, [5 x i32] zeroinitializer, [5 x i32] zeroinitializer, [5 x i32] zeroinitializer], align 16

**Comment by Yawningx** [comment]
解决了，是我的类型解析有问题。😭
